clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java

```diff
@@ -4250,6 +4251,87 @@ public void testBatchOnlyStoreDataRecovery() {
+   * 5. The Consumer thread continues to process the message, but the mismatch in the PCS leader-follower state causes
+   * sanity check failures (e.g. {@link LeaderFollowerStoreIngestionTask#validateRecordBeforeProducingToLocalKafka})
+   * 6. This eventually results in the SIT being killed, which can only be recovered by manually restarting the Venice
+   * server, which results in a massive spike in ingestion latency
+   * <p>
+   * Specifically, this tests the follower to leader transition. Followers should always consume from local VT, but
+   * leaders should never consume from local VT, and an ingestion exception will be emitted from
+   * {@link LeaderFollowerStoreIngestionTask#validateRecordBeforeProducingToLocalKafka} if this happens.
+   * By adding a reader lock to the entire section after {@link LeaderFollowerStoreIngestionTask#shouldProcessRecord},
+   * the PCS state can never be modified while the Consumer thread is processing messages.
+   */
+  @Test(dataProvider = "aaConfigProvider")
+  public void testShouldProcessRecord(AAConfig aaConfig) throws Exception {
+    // Create a batch of polled messages for the consumer thread to process. Only one message is necessary
+    PubSubTopicPartition topicPartition = new PubSubTopicPartitionImpl(pubSubTopic, PARTITION_FOO);
+    PubSubHelper.MutablePubSubMessage pubSubMessage = PubSubHelper.getDummyPubSubMessage(false);
+    pubSubMessage.setTopicPartition(topicPartition);
+    List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> consumedMessages = Arrays.asList(pubSubMessage);
+
+    // Set the RT topic stuff on the PCS along with the leader state to eliminate the scary exceptions due to test setup
+    PubSubTopic rtTopic = pubSubTopicRepository.getTopic(Version.composeRealTimeTopic(storeNameWithoutVersionInfo));
+    TopicSwitch topicSwitch = new TopicSwitch();
+    topicSwitch.sourceKafkaServers = Arrays.asList(inMemoryRemoteKafkaBroker.getKafkaBootstrapServer());
+    topicSwitch.sourceTopicName = rtTopic.getName();
+    TopicSwitchWrapper topicSwitchWrapper =
+        new TopicSwitchWrapper(topicSwitch, pubSubTopicRepository.getTopic(topicSwitch.sourceTopicName.toString()));
+
+    runTest(Collections.singleton(PARTITION_FOO), () -> {
+      TestUtils.waitForNonDeterministicAssertion(
+          5,
+          TimeUnit.SECONDS,
+          () -> assertTrue(storeIngestionTaskUnderTest.hasAnySubscription()));
+
+      VeniceWriter writer = getVeniceWriter(topic, new MockInMemoryProducerAdapter(inMemoryLocalKafkaBroker));
+      doReturn(writer).when(mockWriterFactory).createVeniceWriter(any(VeniceWriterOptions.class)); // prevents NPE
+
+      Thread stateTransitionThread = new Thread(() -> {
+        /*
+         * Wait for the main test thread to go past shouldProcessRecord() and reach waitUntilValueSchemaAvailable()
+         * inside waitReadyToProcessRecord(), then free the main test thread by making the value schema available
+         */
+        Utils.sleep(1000L);
```

Comment:
This time you set here is not quite deterministic in test, not sure if there is a way to make this wait deterministic.

Reply:
Yeah, I would've liked to make the wait way shorter than 1 second, but I can't think of a way to make it deterministic either.

Reply:
Maybe set a flag inside SIT to indicate `waitUntilValueSchemaAvailable` and then wait until that flag is turned on?